core/integration/module_type_test.go

```diff
@@ -1397,9 +1397,31 @@ const (
 	MyEnumNull  MyEnum = "null"
 )
 
-func (m *Dep) Thing(f MyEnum) MyEnum {
+func (m *Dep) Thing(
+	f MyEnum,
+) MyEnum {
 	return f
 }
+
+func (m *Dep) ThingOpt(
+	// +optional
+	// +default="false"
+	f MyEnum,
+) MyEnum {
+	if f == "" {
+		// this is a weird case - if null is passed, it could either be:
+		// - an explicit empty optional value (we choose this one)
+		// - or the enum value "null"
+		return MyEnumNull
+	}
+	return f
+}
```

Comment:
Aha: https://github.com/dagger/dagger/issues/6749#issuecomment-2127757398

> Update: had a discussion with the team and it might be a bug that the engine doesn't accept the null values, so we'll investigate where that comes from and go from there. Seems like passing `null` is something we should allow.

This seems relevant ^

Slightly curious if you remember *what* we discussed and what we thought an explicit `null` value should do.

- e.g. what should `echo '{container{from(address:null)}}' | dagger query` do? Today it errors with:

  ```
  Error: make request: input: container.from resolve: Container.from: missing required argument: "address"
  ```

  I'd like to make it error with `null is not a valid value for string value "address"`

- e.g. what should we do with echo '{container(platform:null){from(address:"alpine")}}' | dagger query`. Today that succeeds, which makes sense - I guess `null` is a valid value for an optional string. But what about if `platform` had a default? Would we want a different behavior there?

Reply:
`null` should be allowed on any non-nullable types. It's up to the resolvers (function that implements a field) to decide what to do with that `null`. I think there's possibly quite a few fields in core that are currently marked as nullable but could be non-null.

I give an example in https://github.com/dagger/dagger/issues/6749. We currently define a description like this:

```go
Description string `default:""`
```
Which produces the following schema:
```graphql
description: String = ""
```

As you can see, in Go it's a non-null, but the schema is produced as nullable. I think we should use a pointer to allow a null there:

```go
Foo string `default:""`
Bar *string `default:""`
```

Which would produce:

```graphql
foo: String! = ""
bar: String = ""
```

>   I'd like to make it error with `null is not a valid value for string value "address"`

Yes, we should error in a general way if a null is being passed to any non-null type, before even calling the resolver.


> * e.g. what should we do with echo `'{container(platform:null){from(address:"alpine")}}' | dagger query`. Today that succeeds, which makes sense - I guess `null` is a valid value for an optional string. But what about if `platform` had a default? Would we want a different behavior there?

The problem is the concept of "optional" here. That `platform` argument is a nullable type. In GraphQL optionals are either nullable types or types that have a default. The reason why is because **GraphQL implicitly uses the default value of `null` for nullable types that don't have a default**.

So if you have this:
```graphql
platform: Platform
```

GraphQL implicitly turns it into:
```graphql
platform: Platform = null
```

That's why nullables without a default are also optional.

So if you omit that value from the query, **GraphQL should send `null`** to the resolver (because of the implicit default).

And if you have a nullable type with a default:

```graphql
foo: String = "bar"
```

It allows `null` be explicitly provided, and only sets "bar" if the argument is **omitted**.

Now how about a non-null with a default value?

```graphql
foo: String! = "bar"
```

Then it doesn't allow a `null` to be provided but it can still be ommitted (still optional), which is when "bar" is set.

In any case, GraphQL should send a value for all field arguments, whether it's the provided value, the default value if not provided, even an implicit `null` if a nullable type is omitted. Otherwise it should error, like sending null to a non-null type.

Reply:
Aha, thanks for the very detailed response :tada:

So I think this clarifies what to do for this PR:
- `null` is allowed in graphql args - only for optional types + enums where `null` is a member.
- Everywhere else it should explicitly error out.
- When it's explicitly specified, it gets directly passed to a resolver, no default arg is applied.

I think that's very doable, though I think changing to pointer receivers in go is *probably* an out of scope change for just now, but I'll try plumbing it through.

Reply:
Can you clarify what you mean by null being a member of an enum?

Reply:
> `null` shouldn't be allowed as a value

So how should this case be handled? https://github.com/dagger/dagger/pull/8929/files#diff-b52311de62cf1719ff09068c8c5738bcdc164046aa259c85fc314d84c378e7bcL1392-R1398

The problem is that values like `true`/`false`/`null` are ambiguous - they can be either graphql enums or literal bools/null values. In the context of a boolean argument, `true` should resolve to the boolean, but in the context of an enum that has it as a member, it should resolve as the enum member.

See https://github.com/dagger/dagger/pull/8682.

Reply:
I see. But from the context of an enum member, that should be interpreted as a string with value `"null"` instead of an actual `null`.

Side node: the convention for GraphQL enum values is `NULL` (SCREAMING_SNAKE_CASE), so the value `null` should only be possible through a custom enum in a module, but we could make that conversion internally (as part of our planned changes on names/values).

Reply:
I suggest changing the casing here if it's registered as "null" by the SDK, at least in this specific case. It would avoid the conflict issue. Otherwise probably a good idea to forbid to avoid issues.

Reply:
Bleh okay. Well in that case we should just explicitly error out.

We should have the auto caps for the enum values, but I think we should do that separately.

Reply:
Updated the title and description of this PR, and pushed a commit to do as suggested.

Let's see how the tests do with this change.